akka-remote/src/main/scala/akka/remote/artery/tcp/SSLEngineProvider.scala

```diff
@@ -0,0 +1,331 @@
+import javax.security.auth.kerberos.KerberosPrincipal
+
+import scala.annotation.tailrec
+import scala.util.Try
+
+import akka.actor.ActorSystem
+import akka.annotation.ApiMayChange
+import akka.annotation.InternalApi
+import akka.event.LogMarker
+import akka.event.Logging
+import akka.event.MarkerLoggingAdapter
+import akka.japi.Util.immutableSeq
+import akka.remote.security.provider.AkkaProvider
+import akka.stream.ConnectionException
+import akka.stream.IgnoreComplete
+import akka.stream.TLSClosing
+import akka.stream.TLSRole
+import com.typesafe.config.Config
+import com.typesafe.sslconfig.Base64
+import sun.security.util.HostnameChecker
+
+@ApiMayChange trait SSLEngineProvider {
+
+  def createServerSSLEngine(hostname: String, port: Int): SSLEngine
+
+  def createClientSSLEngine(hostname: String, port: Int): SSLEngine
+
+  /**
+   * Verification that will be called after every successful handshake
+   * to verify additional session information. Return `None` if valid
+   * otherwise `Some` with explaining cause.
+   */
+  def verifyClientSession(hostname: String, session: SSLSession): Option[Throwable]
+
+  /**
+   * Verification that will be called after every successful handshake
+   * to verify additional session information. Return `None` if valid
+   * otherwise `Some` with explaining cause.
+   */
+  def verifyServerSession(hostname: String, session: SSLSession): Option[Throwable]
+
+}
+
+class SslTransportException(message: String, cause: Throwable) extends RuntimeException(message, cause)
+
+/**
+ * INTERNAL API: only public via config
+ * Config in
+ */
+@InternalApi private[akka] final class ConfigSSLEngineProvider(config: Config, log: MarkerLoggingAdapter) extends SSLEngineProvider {
+
+  def this(system: ActorSystem) = this(
+    system.settings.config.getConfig("akka.remote.artery.ssl.config-ssl-engine"),
+    Logging.withMarker(system, classOf[ConfigSSLEngineProvider].getName))
+
+  private val SSLKeyStore = config.getString("key-store")
+  private val SSLTrustStore = config.getString("trust-store")
+  private val SSLKeyStorePassword = config.getString("key-store-password")
+  private val SSLKeyPassword = config.getString("key-password")
+  private val SSLTrustStorePassword = config.getString("trust-store-password")
+  val SSLEnabledAlgorithms = immutableSeq(config.getStringList("enabled-algorithms")).to[Set]
+  val SSLProtocol = config.getString("protocol")
+  val SSLRandomNumberGenerator = config.getString("random-number-generator")
+  val SSLRequireMutualAuthentication = config.getBoolean("require-mutual-authentication")
+
+  private val hostnameVerifier: HostnameVerifier = config.getBoolean("hostname-verification") match {
+    case false ⇒ new DisabledComplainingHostnameVerifier(log) // FIXME is this a good default or should we have a completely silent default?
+    case true  ⇒ new DefaultHostnameVerifier(log)
+  }
+
+  private val sslContext = new AtomicReference[SSLContext]()
```

Comment:
Could be a `lazy val` I guess? I think the construction in our old `SSLSettings` was to be able to pass in a logger to `getOrCreateContext`.